### PR TITLE
Use futex in concurrent squeue

### PR DESCRIPTION
**Pull request by Kestrer**
Based on my ideas in #46 - now two threads can push to the squeue in parallel, although it does require synchronization in the end.

### CODE REVIEW

File: src/concurrent/squeue.rs [hunk 1]
```diff
@@ -67,18 +64,63 @@ impl SubmissionQueue<'_> {
...
+        while (*self.queue.tail)
+            .compare_exchange(
+                previous_reserved_tail,
+                previous_reserved_tail.wrapping_add(1),
+                atomic::Ordering::Release,
+                atomic::Ordering::Relaxed,
+            )
+            .is_err()
+        {
+            libc::syscall(
```
Comment (quininer) [inline src/concurrent/squeue.rs:98]: I think we shouldn't wait when tail == previous_reserved_tail?

File: src/concurrent/squeue.rs [hunk 2]
```diff
@@ -67,18 +64,63 @@ impl SubmissionQueue<'_> {
...
+                std::ptr::null::<u32>(),
+                // Ignored by FUTEX_WAIt.
+                0_u32,
+            );
         }
 
-        *self.queue.sqes.add((tail & self.ring_mask) as usize) = entry;
-
-        (*self.queue.tail).store(tail.wrapping_add(1), atomic::Ordering::Release);
+        libc::syscall(
```
Comment (quininer) [inline src/concurrent/squeue.rs:112]: We should add a function that does not depend on libc. like https://github.com/tokio-rs/io-uring/blob/master/src/sys/mod.rs#L20

**Comment by quininer** [comment]
We also need to add a bench to prove that it has a performance advantage over locks.

File: src/concurrent/squeue.rs (hunk 1 above)
Comment (Koxiaet) [reply src/concurrent/squeue.rs]: Yes, my bad - I've fixed that now.

**Comment by Koxiaet** [comment]
So I wrote some improved benchmarks to try and see how this would change performance - but to my dismay this change actually _decreases_ performance in the heavily contended case. The lock-based approach took about 130ns to push a single item, whereas the futex approach takes 190ns! I committed the benchmarks I wrote - so before we close this we should at least check them to make sure they were good benchmarks.

**Comment by quininer** [comment]
The result is a bit regretful. I guess this is because it almost always never go to `futex_wait` when heavily contended, so that it degenerates into a spinlock.

**Pull request closed without merging**